pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,438 @@
+// on namespaceUpdateMap and latestFeatureStates
+type pendingCRUpdates struct {
+	// lock to acquire before updating namespaceUpdateMap and latestFeatureStates
+	lock *sync.RWMutex
+	// holds namespace name and status to update CR in the namespace
+	// map value true, means update is pending, false means no update is required.
+	namespaceUpdateMap map[string]bool
+	// latest featureStates
+	latestFeatureStates []featurestatesv1alpha1.FeatureState
+}
+
+var (
+	supervisorFeatureStatConfigMapName       string
+	supervisorFeatureStateConfigMapNamespace string
+	pendingCRUpdatesObj                      *pendingCRUpdates
+	k8sClient                                clientset.Interface
+	controllerRuntimeClient                  client.Client
+)
+
+// StartSvFSSReplicationService Starts SvFSSReplicationService
+func StartSvFSSReplicationService(ctx context.Context, svFeatureStatConfigMapName string, svFeatureStateConfigMapNamespace string) error {
+	log := logger.GetLogger(ctx)
+	log.Info("Starting SvFSSReplicationService")
+
+	supervisorFeatureStatConfigMapName = svFeatureStatConfigMapName
+	supervisorFeatureStateConfigMapNamespace = svFeatureStateConfigMapNamespace
+	pendingCRUpdatesObj = &pendingCRUpdates{
+		lock:                &sync.RWMutex{},
+		namespaceUpdateMap:  make(map[string]bool),
+		latestFeatureStates: make([]featurestatesv1alpha1.FeatureState, 0),
+	}
+	var err error
+	// This is idempotent if CRD is pre-created then we continue with initialization of svFSSReplicationService
+	err = k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		reflect.TypeOf(featurestatesv1alpha1.CnsCsiSvFeatureStates{}).Name(), CRDGroupName, internalapis.SchemeGroupVersion.Version, apiextensionsv1beta1.NamespaceScoped)
+	if err != nil {
+		log.Errorf("failed to create CnsCsiSvFeatureStates CRD. Error: %v", err)
+		return err
+	}
+
+	// Create the kubernetes client
+	k8sClient, err = k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("create k8s client failed. Err: %v", err)
+		return err
+	}
+	// get kube config
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Error: %v", err)
+		return err
+	}
+	// create controller runtime client
+	controllerRuntimeClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("failed to create controllerRuntimeClient. Err: %v", err)
+		return err
+	}
+
+	pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
+	if err != nil {
+		log.Errorf("failed to retrieve latest feature switch states. Error: %v", err)
+		return err
+	}
+	workloadNamespaces, err := getAllWorkloadNamespaces(ctx)
+	if err != nil {
+		log.Errorf("failed to retrieve workload namespaces. Error: %v", err)
+		return err
+	}
+	for _, namespace := range workloadNamespaces.Items {
+		pendingCRUpdatesObj.namespaceUpdateMap[namespace.Name] = true
```

Comment:
Now that you are listening to namespace add event, is this required?

Reply:
We can remove this, as `func namespaceAdded(obj interface{}) {` is taking care of adding workload namespaces to `pendingCRUpdatesObj.namespaceUpdateMap` map.